nft-client/src/cts1.rs

```diff
@@ -0,0 +1,377 @@
+        let len = u8::try_from(self.0.len()).map_err(|_| W::Err::default())?;
+        len.serial(out)?;
+        for transfer in &self.0 {
+            transfer.serial(out)?;
+        }
+        Ok(())
+    }
+}
+
+/// The type of update for an operator update.
+pub enum OperatorUpdate {
+    /// Remove the operator.
+    Remove,
+    /// Add an address as an operator.
+    Add,
+}
+
+/// The deserialization of an operator update as according to the CTS1
+/// specification.
+impl Deserial for OperatorUpdate {
+    fn deserial<R: Read>(source: &mut R) -> Result<Self, ParseError> {
+        let discriminant = source.read_u8()?;
+        match discriminant {
+            0 => Ok(OperatorUpdate::Remove),
+            1 => Ok(OperatorUpdate::Add),
+            _ => Err(ParseError::default()),
+        }
+    }
+}
+
+/// A SHA256 digest.
+pub type Sha256 = [u8; 32];
+
+/// A URL for the metadata.
+pub struct MetadataUrl {
+    /// The url encoded as according to CTS1.
+    pub url:  String,
+    /// An optional checksum of the content found at the URL.
+    pub hash: Option<Sha256>,
+}
+
+/// Deserialization for MetadataUrl as according to the CTS1 specification.
+impl Deserial for MetadataUrl {
+    fn deserial<R: Read>(source: &mut R) -> Result<Self, ParseError> {
+        let len = source.read_u16()?;
+        let mut bytes = Vec::new();
+        for _ in 0..len {
+            bytes.push(source.read_u8()?)
+        }
+        let url = String::from_utf8(bytes)?; //.map_err(|e| ParseError::default())?
+        let hash = Option::<Sha256>::deserial(source)?;
+        Ok(MetadataUrl { url, hash })
+    }
+}
+
+/// Smart contract logged event, part of the CTS1 specification.
+pub enum Event {
+    /// Transfer of an amount of tokens
+    Transfer {
+        token_id: TokenIdVec,
+        amount:   TokenAmount,
+        from:     Address,
+        to:       Address,
+    },
+    /// Minting an amount of tokens
+    Mint {
+        token_id: TokenIdVec,
+        amount:   TokenAmount,
+        owner:    Address,
+    },
+    /// Burning an amount of tokens
+    Burn {
+        token_id: TokenIdVec,
+        amount:   TokenAmount,
+        owner:    Address,
+    },
+    /// Add/Remove an address as operator for some other address.
+    UpdateOperator {
+        update:   OperatorUpdate,
+        owner:    Address,
+        operator: Address,
+    },
+    /// Provide an URL with the metadata for a certain token.
+    TokenMetadata {
+        token_id:     TokenIdVec,
+        metadata_url: MetadataUrl,
+    },
+    /// Custom event outside of the CTS1 specification.
+    Unknown,
+}
+
+impl Display for Event {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        match self {
+            Event::Transfer {
+                token_id,
+                from,
+                to,
+                amount,
+            } => {
+                if *amount > 0 {
```

Comment:
Why the special case? This will print nothing if amount is 0, why is that sensible?

The standard mandates that transfers of 0 are treated the same, right?